### PR TITLE
Introducing filters and host reconstruction

### PR DESCRIPTION
**Pull request by JacobCallahan**
Filters are now available to decide which hosts are selected for certain actions.
VMBroker can now reconstruct hosts from the local inventory

### CODE REVIEW

**Comment by JacobCallahan** [comment]
From the README:

**Filter hosts for Broker actions**
Actions that Broker can take against hosts (checkin, duplicate, extend) can take in a filter argument. This filter will decide which hosts the actions are applied to. A filter by itself will not select hosts for these actions, you will still need to specify which hosts to act against, or use `--all` when available. From there, the filter decides which of those hosts make it through to be acted upon.
Broker's filters are based on what is stored in its local inventory file. As such, only properties in that file are filter-able. Nested properties are annotated with a `.` notation. For example, a top-level property `hostname` can be accessed by itself. However, a nested property of `_broker_args` called `version` would be accessed by `_broker_args.version`.
Filters take the form `"(property)(condition)(value)"`. Filters have several possible conditions:
 - `<` means "in" or that the filter value exists within the actual value
 - `=` means "equals"
 - `{` means "starts with"
 - `}` means "ends with"
Furthermore, putting a `!` before the condition inverts the filter. So `!=` means "not equals" and `!<` means "not in".
**Example filters:**
`--filter 'hostname<test'` The string test should exist somewhere in the hostname value
`--filter '_broker_args.template{deploy-sat'` The template should start with the string "deploy-sat"
You can also chain multiple filters together by separating them with a comma.
`--filter 'name<test,_broker_args.provider!=RHEV'` The host's name should have test in it and the provider should not equal RHEV.
**Note:** Due to shell expansion, it is recommended to wrap a filter in single quotes.

File: README.md [hunk 1]
```diff
@@ -104,3 +109,28 @@ broker checkin -b --all
 broker duplicate -b 0
 broker execute -b --workflow my-awesome-workflow --artifacts
 ```
+
+**Filter hosts for Broker actions**
+
+Actions that Broker can take against hosts (checkin, duplicate, extend) can take in a filter argument. This filter will decide which hosts the actions are applied to. A filter by itself will not select hosts for these actions, you will still need to specify which hosts to act against, or use `--all` when available. From there, the filter decides which of those hosts make it through to be acted upon.
+
+Broker's filters are based on what is stored in its local inventory file. As Therefore, only properties in that file are filter-able. Nested properties are annotated with a `.` notation. For example, a top-level property `hostname` can be accessed by itself. However, a nested property of `_broker_args` called `version` would be accessed by `_broker_args.version`.
```
Comment (swadeley) [inline README.md:117]:
```suggestion
Broker's filters are based on what is stored in its local inventory file. Therefore, only properties in that file are filter-able. Nested properties are annotated with a `.` notation. For example, a top-level property `hostname` can be accessed by itself. However, a nested property of `_broker_args` called `version` would be accessed by `_broker_args.version`.
```

File: README.md [hunk 2]
```diff
@@ -119,3 +124,28 @@ broker checkin -b --all
...
+Furthermore, putting a `!` before the condition inverts the filter. So `!=` means "not equals" and `!<` means "not in".
+
+**Example filters:**
+
+`--filter 'hostname<test'` The string test should exist somewhere in the hostname value
+`--filter '_broker_args.template{deploy-sat'` The template should start with the string "deploy-sat"
+
+You can also chain multiple filters together by separating them with a comma.
+
+`--filter 'name<test,_broker_args.provider!=RHEV'` The host's name should have test in it and the provider should not equal RHEV.
```
Comment (mshriver) [inline README.md:149]: I think this piece should have a bit more explicit documentation around it - that the filters are applied with AND, not OR logic.

File: broker/helpers.py [hunk 3]
```diff
@@ -67,6 +65,44 @@ def flatten_dict(nested_dict, parent_key=""):
...
+            return k, v, test
+
+
+def inventory_filter(inventory, raw_filter):
+    """Filter out inventory items depending on the filter provided"""
+    resolved_filter = classify_filter(raw_filter)
+    matching = []
+    for host in inventory:
+        flattened_host = flatten_dict(host, separator=".")
+        if not isinstance(resolved_filter, list):
```
Comment (mshriver) [inline broker/helpers.py:94]: This doesn't need to be done within every iteration of the loop right?

File: broker/helpers.py [hunk 4]
```diff
@@ -67,6 +65,44 @@ def flatten_dict(nested_dict, parent_key=""):
...
+            resolved_filter = [resolved_filter]
+        eval_list = [
+            eval(rf[2].format(haystack=flattened_host[rf[0]], needle=rf[1]))
```
Comment (mshriver) [inline broker/helpers.py:97]: I think a simple structure like a NamedTuple would make this more readable, instead of just the tuple indexes.

Comment (mshriver) [reply broker/helpers.py:97]: 100% opinion here btw, just a thought

Comment (JacobCallahan) [reply broker/helpers.py]: not a bad idea. i've updated the current version to use a named tuple.

**Review by mshriver — APPROVED** [review]

**Pull request merged**